### ISSUE TITLE
Main modules sometimes build and sometimes don't depending on how you name them

### ISSUE THREAD

**Issue by chshersh** [tomjaguarpaw/tilapia#36]
Executable stanzas in Haskell (executable, test, benchmarks) should have `Main` modules in them. However, Haskell allows naming `Main` modules with a different filename.

For instance, this is a perfectly fine content of the `Main` module in the `test/Spec.hs` for the test suite stanza:

```haskell
module Main (main) where

import Foo (projectName)


main :: IO ()
main = putStrLn ("Tests for " ++ projectName)
```

In tests, it's common to have the `Spec.hs` file for unit tests and the `Doctest.hs` file for documentation tests in the same directory on the same level. And they work and compile fine.

However, the same approach doesn't work for executable stanzas. If I put two `Main` modules inside a single directory under different executable stanzas:

```cabal
executable foo-exe1
  hs-source-dirs:      app
  main-is:             Foo.hs
  build-depends:       base, foo

executable foo-exe2
  hs-source-dirs:      app
  main-is:             Exe.hs
  build-depends:       base, foo
```

Suddenly, I see compilation errors:

```
Preprocessing executable 'foo-exe1' for foo-0.0.0.0..
Building executable 'foo-exe1' for foo-0.0.0.0..
Preprocessing test suite 'foo-test' for foo-0.0.0.0..
Building test suite 'foo-test' for foo-0.0.0.0..
Preprocessing executable 'foo-exe2' for foo-0.0.0.0..
Building executable 'foo-exe2' for foo-0.0.0.0..

app/Foo.hs:1:8: error:
    File name does not match module name:
    Saw: ‘Main’
    Expected: ‘Foo’
  |
1 | module Main (main) where
  |        ^^^^
[1 of 1] Compiling Main

app/Foo.hs:1:8: error:
    File name does not match module name:
    Saw: ‘Main’
    Expected: ‘Foo’
  |
1 | module Main (main) where
  |        ^^^^
```

Moreover, if I rename one file to `Main.hs`, everything starts working again! This can be very confusing and frustrating sometimes.

On a related note, when trying to produce the HIE files for projects (relevant for static analysers and IDEs), GHC produces the same `Main.hie` file for `Main` modules even having different names, so the content is overridden. And this type of inconsistency in behaviour between different parts of GHC also can be confusing sometimes.

* https://github.com/kowainik/stan/issues/321

P.S. Nice project! I hope you don't mind me sharing some of the issues I have. I hope this can improve the UX of Haskell developers :slightly_smiling_face:

**Comment by tomjaguarpaw** [comment]
> However, Haskell allows naming Main modules with a different filename.

Yes, I find this a bit strange. A Haskell file placed at `Foo/Bar/Baz.hs` *must* have module name `Foo.Bar.Baz` *except* it could also have module name `Main`. Sometimes the `Main` module is allowed to omit the `module Main where` declaration  but I haven't determined exactly under which conditions. It seems that `ghc` allows it but `cabal` does not.

> if I rename one file to Main.hs, everything starts working again!

Interesting. Do you mean if you rename one file to `Main.hs` then *both* executable stanzas start working, or only the one that you renamed to `Main.hs`?

> P.S. Nice project! I hope you don't mind me sharing some of the issues I have. I hope this can improve the UX of Haskell developers 

You're welcome! Please keep sharing. Hopefully we can improve things together.

**Comment by chshersh** [comment]
>  Do you mean if you rename one file to Main.hs then both executable stanzas start working, or only the one that you renamed to Main.hs?

Yeah, both executable stanzas start working. Though, I haven't tried with 3 stanzas, where one stanza is `Main.hs` and the other two have different names. So, I'm not sure how this rule works :confused:

**Comment by tomjaguarpaw** [comment]
> > Do you mean if you rename one file to Main.hs then both executable stanzas start working, or only the one that you renamed to Main.hs?

> Yeah, both executable stanzas start working.

This is truly one of the strange and funny mysteries of cabal.
